NetKet/Machine/layer.hpp

```diff
@@ -0,0 +1,130 @@
+// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "abstract_layer.hpp"
+#include "activations.hpp"
+#include "conv_layer.hpp"
+#include "fullconn_layer.hpp"
+#include "sum_output.hpp"
+
+#ifndef NETKET_LAYER_HPP
+#define NETKET_LAYER_HPP
+
+namespace netket {
+template <class T>
+class Layer : public AbstractLayer<T> {
+  using Ptype = std::unique_ptr<AbstractLayer<T>>;
+
+  Ptype m_;
+
+ public:
+  using VectorType = typename AbstractMachine<T>::VectorType;
+  using MatrixType = typename AbstractMachine<T>::MatrixType;
+  using StateType = typename AbstractMachine<T>::StateType;
+  using LookupType = typename AbstractMachine<T>::LookupType;
+
+  explicit Layer(const Graph &graph, const json &pars) { Init(graph, pars); }
+
+  void Init(const Graph &graph, const json &pars) {
+    CheckInput(pars);
+    if (pars["Name"] == "FullyConnected") {
+      if (pars["Activation"] == "Lncosh") {
```

Comment:
To improve readability: Maybe extract the value once as a variable and use that in every if statement (i.e., `const std::string activation = pars["Activation"]`).

Reply:
But not all layers need an activation